### PR TITLE
Include auto-completion for dot params when tagged with @inheritDotParams

### PR DESCRIPTION
**Pull request by orgadish**
### Intent

Addresses #16069

This PR results in Rstudio autocompletion expanding dot params when the function is tagged with roxygen `@inheritDotParams`. 

For example, this is what it now looks like for `ggplot2::scale_color_grey` which inherits dot params from `ggplot2::discrete_scale`:
Image
image

### Approach

Add a method to SessionRCompletions that:
1. Retrieves the help documentation for the function.
2. Checks if `...` exists with the precise formatting associated with `@inheritDotParams`. (Since we can't test for the presence of the roxygen tag directly we have to test for its result: Dot params with the text "arguments passed to" and children items with `\code`.
3. If (2) is true, the referenced function's formals are added to the autocompletions, except where a local one already exists, or if that function has `...`. Since the goal of this is to help with simple wrappers, I decided to force it to be shallow -- it does not recurse.

Note: This PR was developed in concert with Claude.

### Automated Tests

Tests for this capability are in a new `test-completions-inherit-dot-params.R` file. 

```r
> testthat::test_dir("src/cpp/tests/testthat", filter = "completion")
✔ | F W  S  OK | Context
✔ |         31 | auto-completion                                                                       
✔ |         26 | inheritDotParams completions                                                          

══ Results ════════════════════════════════════════════════════════════════════════════════════════════
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 57 ]
```

_Note: Since `rstudio` is not a typical R package that can leverage tools like `devtools::load_all()`, I ran the test locally by updating the `SessionRCompletions.R` file in my local RStudio installation._

#### Steps to Test Manually:
Load RStudio with the updated `SessionRCompletions.R` file in `src/cpp/session/modules`.

Test autocompletion on a function that uses `@inheritDotParams`, such as `ggplot2::scale_color_grey` (which uses `@inheritDotParams discrete_scale -expand -position -scale_name -palette -fallback.palette`).

1. In the console, type `ggplot2::scale_color_grey(` and press tab. You should see all the the formals of `scale_color_grey` (including `...`) AND the inherited formals from `discrete_scale`, labeled as such. You You should NOT see the excluded items from `discrete_scale` in the list above, e.g. `expand`, `position`, etc.
2. Test without package name prefix, by loading `ggplot2` (e.g. running `library(ggplot2)`) and then typing `scale_color_grey(` and pressing tab to trigger auto-complete. You should now see the same thing as before but rhs do not have the package name.
3. Test with a partial token for inherited args, for example, by typing `scale_color_grey(br`. Autocomplete (either by waiting or pressing tab) should show `breaks` which comes from `discrete_scale` and not from scale_color_grey's formals.
5. Test with a partial token for standard formals, for example, by typing `scale_color_grey(st` and you should see `start`.
6. Test on a function that has dots but not `@inheritDotParams`, such as by typing `scale_alpha` and triggering autocomplete via tab. That should show the formals for scale_alpha as normal, including the dots.

#### Testing Results with screenshots 
All tests above pass

1. image
2. image
3. image
4. image
5. image

##### Performance
Performing this completion analysis requires an extra call to `utils::help` (via `.rs.makeHelpCall`) and `utils:::.getHelpFile`. The worst I have seen running `.rs.getCompletionsFunction` on `ggplot2::scale_color_grey` for the first time, it took 1.77s, presumably to load all the help for ggplot2. But afterwards, that is cached and the calls were instantaneous. In that case, however, RStudio felt especially laggy so there may have been other issues with my computer. In subsequent runs, where RStudio did not feel laggy, the pre-cached time was 0.02-0.06s, so I believe this is low risk.

##### Changing existing behavior of `@inheritDotParams`
The approach for this PR takes advantage of the existing `@inheritDotParams`. This means a forced change in behavior for those that had opted in to what `@inheritDotParams`. This will not cause any errors, but may clutter autocomplete for those functions where not intended. Furthermore, I believe this has been viewed as a limitation of inheritDotParams not the design, and so I think this is a low risk change.

##### Auto-complete is based documentation not the installed version
If a developer uses `@inheritDotParams` with an external package, the help documentation will be based on the version used when the documentation was built, not the one that the end user has currently installed. This is already a limitation of `@inheritDotParams` and is why using it with functions from external packages should only be done with caution.

This new functionality continues that same approach, assuming the documentation is the state of truth. 

TBD

### Documentation

> Specify which documentation has been added or modified and why (User Guide? Admin Guide?). If no documentation was added for a new feature, indicate why. If documentation was added in a separate PR, link the PR here.

I don't believe this requires any RStudio documentation updates, but requires updating roxygen documentation for `@inheritDotParams`.

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md`
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

**Comment by gtritchie** [comment]
Thanks for the contribution! Before we take a look, have you submitted the contributer agreement as outlined here? https://github.com/rstudio/rstudio/blob/main/CONTRIBUTING.md#contributing-code

**Comment by orgadish** [comment]
@gtritchie Yes I have. Sent to contribute@posit.co on Mar 2

**Comment by gtritchie** [comment]
## PR Review Summary

Thanks for the contribution! The feature is well-designed with clean separation between Rd parsing, help-db lookup, and the orchestrator. The test coverage is solid with layered unit tests and real-world end-to-end validation against ggplot2. A few items to address before merging:

---

### Critical (must fix)

**1. `parsed` → `parsedTargetName` rename breaks Shiny completions**

The PR renames `parsed` to `parsedTargetName` in two unrelated Shiny completion functions (~line 3850 in `shinyUICompletions` and ~line 3970 in `shinyServerCompletions`), but the subsequent references on following lines (`if (is.null(parsed))`, `lapply(parsed, ...)`) are **not** updated. This will cause a runtime error (`object 'parsed' not found`) whenever Shiny UI or Server completions are triggered.

These renames appear to be an accidental find-and-replace side effect — the variable stores `parse(file)` output, not a "target name." Please revert both renames.

---

### Important (should fix)

**2. Incorrect return-type docstring on `parseInheritDotParamsFromRd`**

The comment says it returns "a character vector of argument names" but it actually returns `list(targetName = ..., argNames = ...)`. Suggest:

```r
# Returns a list with fields `targetName` (character scalar) and `argNames`
# (character vector), or NULL if the @inheritDotParams pattern is not found.
```

**3. Stale "cache key" comment in `getCompletionsInheritDotParams`**

> This ensures the cache key is always fully qualified, so a user changing the last loaded library cannot cause a stale hit.

There is no caching mechanism in this function or its callees. This appears left over from an earlier design iteration. Suggest removing the sentence entirely, or replacing with:

```r
# If no package was specified via ::, resolve it from the function's
# environment so the correct help database is queried.
```

**4. Inaccurate claim about R help database caching in `parseInheritDotParamsFromHelp`**

> R already caches the help database in memory after first access, so no additional caching is needed here.

Tracing through `utils::help` → `readRDS(aliases.rds)` and `.getHelpFile` → `lazyLoadDBexec` → `readRDS(mapfile)`, R re-reads from disk each time — there is no R-level in-memory cache. The OS filesystem cache may help, but the claim as stated is inaccurate and could discourage a maintainer from adding caching if profiling reveals a bottleneck. Suggest removing or replacing with an accurate note.

**5. Broad `tryCatch` in `parseInheritDotParamsFromHelp` swallows all errors**

The single `tryCatch` wraps both the help lookup (expected to fail for unknown functions) and the internal `utils:::.getHelpFile` call (should not fail unless R internals change). If `.getHelpFile` is renamed/removed in a future R version, the feature will silently stop working with zero diagnostic output. Consider splitting into two `tryCatch` blocks and logging unexpected errors.

---

### Suggestions (nice to have)

**6. Multi-target `@inheritDotParams` mismatch between Loop 1 and Loop 2**

Loop 1 in `parseInheritDotParamsFromRd` searches **forward** and finds the *first* target's sentinel text, but Loop 2 searches in **reverse** and finds the *last* `\describe` block. If multiple `@inheritDotParams` tags exist, the displayed target name and expanded arguments could come from *different* functions. The "shallow" comment should explicitly document this limitation.

**7. Inaccurate claim about primitives**

> Non-package functions (global, anonymous, primitive) cannot have a compiled help database entry

Primitives like `sum`, `c`, `paste` do have help entries in `base`. The code behavior is correct (primitives exit early via the `!nzchar(envName)` guard), but the stated reason is wrong. Suggest:

```r
# Functions whose environment does not resolve to a named package
# namespace (global env, empty env, or missing) are unlikely to have
# @inheritDotParams documentation, so exit early.
```

**8. Missing test coverage for several code paths:**
- Bare function name resolution (all orchestrator tests use `"pkg::wrapper"` — the `environmentName(environment(object))` fallback path is untested)
- `hasDots` fallback logic (when token filtering excludes `"... = "` from `formals$formals`)
- Triple-colon `pkg:::fn` syntax (the regex `:{2,3}` handles it but no test exercises it)
- Backtick-quoted argument names (the code uses `deparse(as.name(fml), backtick = TRUE)` but no test covers unusual names)

**9. Copyright year**

Test file header says `Copyright (C) 2022` — this is new code from 2026.

---

### Strengths

- Clean three-layer architecture (Rd parsing → help lookup → orchestrator) with each layer independently testable
- Solid fixture-based Rd testing with both full and partial `@inheritDotParams` cases
- Correct own-formals and used-args exclusion logic
- Good end-to-end tests with ggplot2 and proper `skip_if_not_installed` guards
- `with_fixture_completions` helper with proper `on.exit` cleanup

**Comment by orgadish** [comment]
Thanks for the clear and thorough review! (It seems I left quite a few things from previous design iterations -- sorry!)

I will let you know when the changes are complete.

**Comment by orgadish** [comment]
@gtritchie I believe I have addressed all the points you made:

1. Reverted accidental name changes to usages of `parsed` in existing code.

2. Updated the return description for `parseInheritDotParamsFromRd` as suggested.

3. Updated the stale comment about cache key, as suggested (caching was implemented in one iteration, but it added significant complexity for little gain -- at least in how I was implementing it). 

4. Removed the comment about caching in `parseInheritDotParamsFromHelp`.

5. Separated `parseInheritDotParamsFromHelp` into 2 `tryCatch` calls -- one of which is expected to error and the other of which presents a `warning`.

6. Updated the comment in `parseInheritDotParamsFromRd` to explicitly point out that we search forward in loop 1 and in reverse in loop 2 and that this could mean different behavior if 2 `@inheritDotParams`, but that this is acceptable given the rarity of this behavior.

7. Updated the comment about primitives, per your suggestion.

8. Added tests to cover all 4 paths previously missing, per your suggestion.

9. Updated the copyright year.


### Re caching

Do you think it's worth trying to implement some kind of caching? I was looking into updating/using `getCompletionsHelp` but currently it only caches the aliases. And  I was trying to avoid touching existing code as much as possible. My tests suggest that after the first call, it's instantaneous (presumably because of caching that RStudio/OS performs), but I don't know what this would be like on a slower file system (e.g. over the network).

**Comment by gtritchie** [comment]
# PR #17149 Review: @inheritDotParams Autocompletion

**Files changed**: 5 | **+591 / -0**

---

## Critical Issues (2 found)

### 1. No caching of help database lookups — performance risk on every keystroke
`SessionRCompletions.R:1088-1096`

Every tab-completion for **any function with `...`** now triggers `utils::help()` + `utils:::.getHelpFile()` + Rd parsing. Most R functions with `...` (e.g., `c()`, `list()`, `paste()`, `mean()`) do **not** use `@inheritDotParams`, so this work is wasted for the vast majority of completions.

The PR author's own measurements show **20–60ms cached, up to 1.77s uncached**. On a hot completion path triggered by every keystroke, even 20ms per invocation adds perceptible lag.

**Recommendation**: Add a memoization cache keyed by `pkg::fnName` that stores both positive and negative results. A simple `new.env(parent = emptyenv())` cache would eliminate repeated lookups:

```r
.rs.inheritDotParamsCache <- new.env(parent = emptyenv())

# In getCompletionsInheritDotParams, before parsing:
cacheKey <- paste(pkgName, fnName, sep = "::")
if (exists(cacheKey, envir = .rs.inheritDotParamsCache))
   return(get(cacheKey, envir = .rs.inheritDotParamsCache))
# ... do the work, then cache the result before returning
```

### 2. Uncaught errors in new code can break ALL completions for a function
`SessionRCompletions.R:1093-1096`

The call to `.rs.getCompletionsInheritDotParams()` from the main `getCompletionsFunction` handler is **not wrapped in tryCatch**. If an unexpected error escapes the inner error handlers, the entire completion request fails — not just the inherited params portion.

```r
# Current — no protection:
result <- .rs.appendCompletions(
   result,
   .rs.getCompletionsInheritDotParams(token, string, object, matchedCall, formals$formals)
)
```

**Recommendation**: Wrap with a defensive tryCatch so the new feature cannot regress existing completions:

```r
inheritedCompletions <- tryCatch(
   .rs.getCompletionsInheritDotParams(token, string, object, matchedCall, formals$formals),
   error = function(e) .rs.emptyCompletions()
)
result <- .rs.appendCompletions(result, inheritedCompletions)
```

---

## Important Issues (4 found)

### 3. Warning message in `parseInheritDotParamsFromHelp` has wrong function name
`SessionRCompletions.R:1307-1314`

The warning says `"getCompletionsInheritDotParams: failed to load Rd"` but is in `parseInheritDotParamsFromHelp`. It also omits the package name, making it hard to diagnose.

### 4. Filter by token before deparsing all arg names
`SessionRCompletions.R:1161-1172`

The code deparses every inherited argument into `"name = "` format, then filters. For functions with 20+ inherited args and a specific token prefix, this is wasteful. Filter by `token` against raw `argNames` first, then deparse only the matches.

### 5. Comment about multiple `@inheritDotParams` is misleading
`SessionRCompletions.R:1237-1240`

The comment says Loop 1 and Loop 2 "will return results from separate blocks." In reality, both loops operate on the same `dotsNode` — the outer loop breaks after finding the first `\item{...}{}` node. Multiple `@inheritDotParams` tags produce separate `\item{...}{}` nodes, so only the first is ever processed. The comment should be updated to accurately describe this behavior.

### 6. Inconsistent error handling: some tryCatch blocks warn, most silently return NULL
Multiple locations

Five `tryCatch(..., error = function(e) NULL/FALSE/"")` blocks silently swallow errors. Only the `.getHelpFile` block uses `warning()`. While silent returns are fine for expected failures (function not found), unexpected errors (corrupted Rd, filesystem issues) should be logged for debuggability. Consider adding warnings for unexpected failure paths, or at minimum the outer defensive tryCatch recommended in issue #2.

---

## Suggestions (4 found)

### 7. Add tests for malformed Rd structures

No tests for malformed Rd (missing child nodes, sentinel text without `\code` node, empty `\describe` blocks). These are the most likely real-world failure scenarios.

### 8. Add test for multiple `@inheritDotParams` targets

The implementation explicitly notes this as a limitation but doesn't test the actual behavior. A test would document and lock in whatever the current behavior is.

### 9. Add test for global env functions returning empty

The early exit for `R_GlobalEnv` / `R_EmptyEnv` functions (line 1138) is untested. Users defining wrapper functions locally would hit this path.

### 10. Validate package name from `environmentName()`
`SessionRCompletions.R:1134-1141`

The resolved `envName` could theoretically be something unexpected. Consider checking `envName %in% loadedNamespaces()` before using it as a package name.

---

## Strengths

- **Excellent architecture**: Clean separation into three layers (Rd parsing, help lookup, orchestration) makes the code testable and maintainable
- **Strong test suite**: 315 lines of well-organized behavioral tests covering happy paths, partial `@inheritDotParams`, reserved words, triple-colon syntax, token filtering, and already-used arg exclusion
- **Realistic fixtures**: The `.Rd` test fixtures accurately replicate roxygen2 output
- **Good documentation**: Function headers explain what, why, and key design decisions
- **Thoughtful design**: Shallow expansion only (no recursion), own formals take precedence, respects partial `@inheritDotParams` exclusions

---

## Recommended Action

1. **Must fix before merge**: Add caching (#1) and defensive tryCatch at integration point (#2)
2. **Should fix**: Correct the warning message (#3), add token-first filtering (#4), fix the misleading comment (#5)
3. **Consider**: Add malformed Rd tests (#7), multiple-target test (#8)
4. **After fixes**: Re-run review to verify

**Comment by orgadish** [comment]
23fd5e7 Updated the PR to address your comments.

1. Caching: My previous statements on timing were misunderstood -- the issue was not 20-60ms cached vs 1.77s uncached. Rather the worst case uncached was 1.77s but most uncached cases were 20-60ms. When cached, system.time displayed all 0s. Nevertheless, I still thought it was worth pursuing your recommended caching approach and implemented that and tested timing in more detail. When RStudio is first opened, the call takes 10-560ms (most commonly ~30ms). This is the equivalent of the 20-60ms and 1.77s I referenced before. Calls that rely on the new caching take ~250us. Calls that only rely on RStudio/R's standard caching take ~5ms. Caching definitely makes a difference

<details>
<summary><i>Detailed timing results of cached vs uncached</i></summary>

```r
# Timing: first-ever cold, subsequent cold, and warm for @inheritDotParams completions

# First-ever call: must happen on a fresh R session and before anything else so R's internal caches
# (help db, namespace loading) are included in the measurement.
get_completion <- function() {
   .rs.getCompletionsInheritDotParams(
      "", "ggplot2::scale_color_grey", ggplot2::scale_color_grey,
      quote(ggplot2::scale_color_grey()), c("na.value = ", "aesthetics = ")
   )
}
first <- system.time(get_completion())

# Setup for the remaining tests
cache  <- .rs.getVar("inheritDotParamsCache")
clear_cache <- function()
   rm(list = ls(envir = cache, all.names = TRUE), envir = cache)

# Cold: cache cleared between each iteration via setup
cold <- microbenchmark::microbenchmark(
   cold = get_completion(),
   setup  = clear_cache(),
   times  = 100
)

# Warm: cache already populated from cold benchmark
warm <- microbenchmark::microbenchmark(
   warm = get_completion(),
   times = 100
)

cat(sprintf("First-ever call: %.1f ms\n\n", first["elapsed"] * 1000))
#> First-ever call: 30.0 ms (fastest was 10ms -> slowest was 560ms)

print(cold)
#> Unit: milliseconds
#> expr    min      lq     mean median      uq     max neval
#> cold 4.4972 5.20055 8.927915 5.7459 6.73145 86.9101   100


print(warm)
#> Unit: microseconds
#> expr   min    lq    mean median    uq   max neval
#> warm 237.8 243.6 257.574  246.9 253.5 821.1   100

```
</details>



2. Added a defensive `tryCatch` to make sure the new functionality can't break existing completion calls.

3. Updating warning message.

4. Update to filter by token before deparsing and converting to "name = ".

5. Updated Loop 1 and Loop 2 to both search in the forward direction. I think the slight performance benefit of searching from the back is outweighed by the confusion and complexity that this causes.

6. Ensured that all `tryCatch` blocks warn and/or have comments explaining why they catch silently.

7-9. Added tests for malformed Rd, multiple `@inheritDotParams`, local functions.

10. Used `envName %in% loadedNamespaces()` to validate the envName and simplify other existing checks.

I also found other paths that weren't previously tested and added tests for those.

All tests pass

```r
> testthat::test_file("src/cpp/tests/testthat/test-completions-inherit-dot-params.R")
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 42 ]
> testthat::test_dir("src/cpp/tests/testthat", filter = "completion")
✔ | F W  S  OK | Context
✔ |         31 | auto-completion                                                                                           
✔ |         42 | inheritDotParams completions                                                                              

══ Results ════════════════════════════════════════════════════════════════════════════════════════════════════════════════
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 73 ]

**Comment by orgadish** [comment]
@gtritchie I believe I have addressed all your recent review recommendations (see above).

@gtritchie @kevinushey What are your thoughts about the fact that this changes the behavior for the existing `@inheritDotParams`, rather than relying on a new decorator? On the one hand, it's nice that it fits in where it makes sense to; on the other hand, it means that this behavior will be added to anyone already using this.

File: src/cpp/session/modules/SessionRCompletions.R [hunk 1]
```diff
@@ -1079,10 +1079,274 @@ assign(x = ".rs.acCompletionTypes",
...
+   
    result
    
 })
 
+# Cache for @inheritDotParams Rd lookups, keyed by "pkg::fn".
+# Stores both positive results (list with targetName/argNames) and negative
+# results (NULL) so that functions without @inheritDotParams are not re-parsed
+# on every keystroke.
+.rs.setVar("inheritDotParamsCache", new.env(parent = emptyenv()))
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: Nit: we usually keep variable definitions like this near the top of the file.

File: src/cpp/session/modules/SessionRCompletions.R [hunk 2]
```diff
@@ -1079,10 +1079,274 @@ assign(x = ".rs.acCompletionTypes",
...
    result <- .rs.appendCompletions(result, ggplotCompletions)
    
+   # If the function has `...` in its formals and the package author used
+   # @inheritDotParams, expand those documented dot-arguments into named
+   # completions, as if they were native arguments of the wrapper.
+   # NOTE: formals$formals is token-filtered by resolveFormals and values are
+   # formatted as "name = ", so check for "... = " first (fast path), then
+   # fall back to formals(object) if the token filtered it out.
+   if ("... = " %in% formals$formals ||
+       tryCatch("..." %in% names(formals(object)), error = function(e) FALSE))
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: Do we need both checks here? Is there ever a case where "... =" could be in formals$formals but _not_ in the formals of the actual object here?

File: src/cpp/session/modules/SessionRCompletions.R [hunk 3]
```diff
@@ -1079,10 +1079,274 @@ assign(x = ".rs.acCompletionTypes",
...
+# Local (explicitly declared) formals always take precedence: any name already
+# present in the wrapper's own formals is excluded from the returned set,
+# matching the behaviour of @inheritParams. Only the first `...` argument is
+# expanded (shallow expansion — we do not recurse into the target function's
+# own `...`).
+.rs.addFunction("getCompletionsInheritDotParams", function(token,
+                                                            string,
+                                                            object,
+                                                            matchedCall,
+                                                            ownFormals)
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: nit: mismatched indent

File: src/cpp/tests/testthat/test-completions-inherit-dot-params.R [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+      ".rs.parseInheritDotParamsFromHelp",
+      function(fnName, pkgName) .rs.parseInheritDotParamsFromRd(fixture_rd),
+      envir = rsEnv
+   )
+
+   force(expr)
+}
+
+test_that("getCompletionsInheritDotParams returns ARGUMENT completions from fixture", {
+   skip_if(!new_functions_loaded, "Source SessionRCompletions.R first")
```
Comment (kevinushey) [inline src/cpp/tests/testthat/test-completions-inherit-dot-params.R]: This skip (and others) seem unnecessary? I assume this was here just for interactive testing?

File: src/cpp/session/modules/SessionRCompletions.R [hunk 5]
```diff
@@ -1079,10 +1079,274 @@ assign(x = ".rs.acCompletionTypes",
...
+   # TEXT node in the description. If it's not there, this is a regular `...`
+   # argument and we have nothing to expand.
+   #
+   # Note: if multiple @inheritDotParams tags exist, only the first block is
+   # used. Multiple @inheritDotParams on a single function is uncommon and
+   # this limitation is accepted.
+
+   # Loop 1: find the sentinel TEXT and extract the target function name from
+   # the \code node immediately following it.
+   inheritDotParamsText <- "Arguments passed on to"
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R:1290]: My only complaint here is that this parsing would fail if roxygen2 were to ever change how it formatted this text. I assume this is coming from:

https://github.com/r-lib/roxygen2/blob/93b87787e9f10e1e451f638224b1a91aa75ce8ac/R/rd-inherit.R#L281-L290

We might want to file an issue at https://github.com/r-lib/roxygen2/issues to indicate that RStudio relies on this behavior, or to make it possible for us to parse this in a more deterministic way.

**Comment by kevinushey** [comment]
> What are your thoughts about the fact that this changes the behavior for the existing @inheritDotParams, rather than relying on a new decorator? On the one hand, it's nice that it fits in where it makes sense to; on the other hand, it means that this behavior will be added to anyone already using this.

I think this is fine -- I think the majority of users will find this behavior very useful as opposed to an impediment. If we get some negative feedback in the future, we could re-evaluate, but that would surprise me.

File: src/cpp/session/modules/SessionRCompletions.R (hunk 2 above)
Comment (orgadish) [reply src/cpp/session/modules/SessionRCompletions.R]: @kevinushey The reason I did this is that `formals$formals` is a fast check. If it's true, we don't need to check `formals(object)`. But when a non-empty token without dots is used, the dots will not be in the filtered formals$formals, so the former would be false while the latter could still be true.

But perhaps that minor performance boost isn't worth the added complexity?

File: src/cpp/session/modules/SessionRCompletions.R (hunk 5 above)
Comment (orgadish) [reply src/cpp/session/modules/SessionRCompletions.R:1290]: Filed r-lib/roxygen2#1808 and will add that as a TODO in the comments.

File: src/cpp/session/modules/SessionRCompletions.R (hunk 2 above)
Comment (kevinushey) [reply src/cpp/session/modules/SessionRCompletions.R]: In practice this is going to be imperceptible, but it's good to know there are potential cases worth guarding against. Let's keep the code as is; thanks for explaining!

**Comment by orgadish** [comment]
@kevinushey Updated the PR with your suggested changes, plus:

1. Added comments with TODOs referencing r-lib/roxygen2#1808 that I filed.
2. Added an `is.primitive` check outside the entire `getCompletionsInheritDotParams` logic to avoid unnecessarily entering, even with the cache. Added tests for this path.


```r
> testthat::test_dir("src/cpp/tests/testthat", filter = "completion")
✔ | F W  S  OK | Context
✔ |         31 | auto-completion                                                                                   
✔ |         44 | inheritDotParams completions [3.7s]                                                               

══ Results ════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 4.2 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 75 ]

File: src/cpp/session/modules/SessionRCompletions.R (hunk 5 above)
Comment (orgadish) [reply src/cpp/session/modules/SessionRCompletions.R:1290]: Added comment in 66d80ba

File: src/cpp/tests/testthat/test-completions-inherit-dot-params.R (hunk 4 above)
Comment (orgadish) [reply src/cpp/tests/testthat/test-completions-inherit-dot-params.R]: Removed in 66d80ba

File: src/cpp/session/modules/SessionRCompletions.R (hunk 3 above)
Comment (orgadish) [reply src/cpp/session/modules/SessionRCompletions.R]: Fixed in 66d80ba

File: src/cpp/session/modules/SessionRCompletions.R (hunk 1 above)
Comment (orgadish) [reply src/cpp/session/modules/SessionRCompletions.R]: Moved to top in 66d80ba

**Comment by kevinushey** [comment]
Merged! Thanks so much for taking the time to prepare a PR; this is a great quality of life feature to have in RStudio.

**Comment by orgadish** [comment]
Thanks @kevinushey.

The template for the PR said I should update the QA Notes on the original ticket. What exactly does that mean? How should I update that ticket besides just closing it?

**Comment by kevinushey** [comment]
After a PR is merged, we keep the issue open so QA can test using the same issue, but move the issue to "Test" to indicate it's ready for testing (I can do that part).

The QA notes are just a general test plan as to how to verify the feature works as expected, and any other areas where testing might be warranted (e.g. because they use shared code paths that were modified.)

I see you included those in your original PR description; it'd be helpful if you copied those over into a new comment on the issue thread at https://github.com/rstudio/rstudio/issues/16069. Thanks!

**Comment by orgadish** [comment]
QA Notes have been copied to the ticket (and updated to account for the caching in the latest implementation). Excited to have this feature live!
